application/btsieve/src/ethereum/queries/event.rs

```diff
@@ -114,49 +113,94 @@ impl QueryType for EventQuery {
     }
 }
 
-impl ShouldExpand for EventQuery {
-    fn should_expand(params: &QueryParams) -> bool {
-        params.expand_results
-    }
+#[derive(Deserialize, Derivative, Debug)]
+#[derivative(Default)]
+#[serde(rename_all = "snake_case")]
+pub enum ReturnAs {
+    #[derivative(Default)]
+    TransactionId,
+    Transaction,
+    Receipt,
+    TransactionAndReceipt,
 }
 
-impl ExpandResult for EventQuery {
+impl ToHttpPayload<EventQuery, ReturnAs> for QueryResult {
     type Client = Web3<Http>;
-    type Item = Transaction;
-
-    // TODO: return TransactionReceipt and not Transaction.
-    // Temporarily return the transaction and not the transaction receipt as the
-    // secret is currently only available in the transaction call data but not
-    // in the receipt. This needs to be fixed with https://github.com/comit-network/comit-rs/issues/638
-    fn expand_result(
-        result: &QueryResult,
-        client: Arc<Web3<Http>>,
+    type Item = PayloadKind;
+
+    fn to_http_payload(
+        &self,
+        return_as: &ReturnAs,
+        client: &Web3<Http>,
     ) -> Result<Vec<Self::Item>, Error> {
-        let futures: Vec<_> = result
-            .0
-            .iter()
-            .filter_map(|tx_id| match hex::decode(clean_0x(tx_id)) {
-                Ok(bytes) => Some(bytes),
-                Err(e) => {
-                    warn!("Skipping {} because it is not valid hex: {:?}", tx_id, e);
-                    None
-                }
-            })
-            .map(|id| {
-                client
-                    .eth()
-                    .transaction(TransactionId::Hash(H256::from_slice(id.as_ref())))
-                    .map_err(Error::Web3)
-            })
-            .collect();
+        let to_payload = |transaction_id: H256| to_payload(client, transaction_id, return_as);
 
-        stream::futures_ordered(futures)
+        self.0
+            .iter()
+            .filter_map(to_h256)
+            .map(to_payload)
+            .collect::<FuturesOrdered<_>>()
             .filter_map(|item| item)
             .collect()
             .wait()
     }
 }
 
+fn to_payload(
+    client: &Web3<Http>,
+    transaction_id: H256,
+    return_as: &ReturnAs,
+) -> Box<dyn Future<Item = Option<PayloadKind>, Error = Error>> {
+    let transaction_future = client
```

Comment:
:+1: This works because these are evaluated lazily, right?

Reply:
Yes correct.